### PR TITLE
Expose docker API to host

### PR DESCRIPTION
**Pull request by praveenkumar**
This is only tested on the mac. I will test it on the Linux also but we need someone who can test this windows. I think on windows `unix:///` doesn't work, we do need to check if named pipe can be used.

Some of the commits can be removed once the forked repo not needed.

### CODE REVIEW

File: go.mod
Comment (cfergeau) [inline go.mod:19]: Can we use this new version with the current bundles, or do we need an updated gvisor-tap-vsock image?

File: cmd/crc/cmd/daemon_linux.go
Comment (cfergeau) [inline cmd/crc/cmd/daemon_linux.go:12]: fwiw, it probably makes more sense to have this commit first in the PR.

File: pkg/crc/machine/vsock.go
Comment (cfergeau) [inline pkg/crc/machine/vsock.go:120]: The other entries list `Protocol` first, better to be consistent. This could be merged with the `append` call right before it.

**Comment by praveenkumar** [comment]
/hold

https://github.com/code-ready/crc/pull/3002#discussion_r804565207

File: go.mod
Comment (praveenkumar) [reply go.mod:19]: @cfergeau we can use current bundle with new version, there is no change in vm command which is part of gvisor-tap-vsock but we do need to update the image sooner than later. This is not blocker for thsi PR.

File: cmd/crc/cmd/daemon_linux.go
Comment (praveenkumar) [reply cmd/crc/cmd/daemon_linux.go:12]: If there is no strong opinion then I would have it as it is.

File: go.mod
Comment (cfergeau) [reply go.mod]: We do need to update the image for this PR? or is this just a general comment about bundle quality?

File: pkg/crc/machine/vsock.go
Comment (cfergeau) [reply pkg/crc/machine/vsock.go:118]: This is only half resolved.

**Review by cfergeau — COMMENTED** [review]
Looks good, still one pending comment, I suspect Windows support will be more work.

File: pkg/crc/machine/vsock.go
Comment (praveenkumar) [reply pkg/crc/machine/vsock.go:117]: Resolved.

**Comment by cfergeau** [comment]
Let's address https://github.com/code-ready/snc/pull/506#discussion_r804736999 before merging this could cause changes in this PR.

File: pkg/crc/machine/vsock.go
Comment (mdlayher) [reply pkg/crc/machine/vsock.go:87]: Same throughout the rest of this function and possibly elsewhere.

Comment (praveenkumar) [reply pkg/crc/machine/vsock.go:87]: @mdlayher Thanks for the tip, will update it as part of new commit message.

File: go.mod
Comment (cfergeau) [reply go.mod]: These snc changes are no longer needed, marking this as resolved.

File: go.mod
Comment (cfergeau) [inline go.mod:82]: The addition of `github.com/mdlayher/vsock v1.1.0` in go.mod was done in the commit updating `gvisor-tap-vsock`, but this first commit is not buildable without this change. I'd squash both commits, mentioning in the commit log that we can drop our fork of mdlayher/vsock since it has the API we need in the version we are switching to.
I've tested vsock socket activation of the crc daemon with v1.1.0 and it works as expected.

File: pkg/crc/machine/vsock.go
Comment (cfergeau) [inline pkg/crc/machine/vsock.go:121]: This should be using net.JoinHostPort as well

Comment (mdlayher) [reply pkg/crc/machine/vsock.go:121]: I believe you're right. I would also recommend constructing a https://pkg.go.dev/net/url#URL and setting its fields as necessary then stringifying at API boundaries. It'll handle all the weird edge cases.

File: pkg/crc/machine/vsock.go
Comment (mdlayher) [reply pkg/crc/machine/vsock.go]: No problem!

File: pkg/crc/machine/vsock.go
Comment (cfergeau) [inline pkg/crc/machine/vsock.go:138]: better to run this through https://pkg.go.dev/net/url#QueryEscape, hopefully gvisor-tap-vsock accepts this.
A bit odd to introduce the fmt.Sprintf() in one commit, to change it to using url.URL in a commit which is not really related to this.

File: pkg/crc/machine/vsock.go
Comment (cfergeau) [inline pkg/crc/machine/vsock.go:140]: no, you want `RawQuery: url.QueryEscape(fmt.Sprintf("key=%s, ...))`, and then `return url.String()`. Or is it `RawQuery: fmt.Sprintf("key=%s", url.QueryEscape(constants.GetPrivateKeyPath())`?

Comment (praveenkumar) [reply pkg/crc/machine/vsock.go:140]: Ah my bad I misread `QueryEscape escapes the string so it can be safely placed inside a URL query. ` this one :(

**Comment by anjannath** [comment]
Tried it on windows, currently it fails during `crc start` with the following error:
```
failed to expose port C:\Users\anath\.crc\machines\crc\docker.sock -> ssh-tunnel://core@192.168.127.2:22/run/podman/podman.sock?key%3DC%3A%5CUsers%5Canath%5C.crc%5Cmachines%5Ccrc%5Cid_ecdsa: key not provided for unix-ssh connection
```
Daemon logs:
```
PS C:\Users\anath\redhat\crc> crc daemon
INFO Checking if running in a shell with administrator rights
INFO Checking Windows 10 release
INFO Checking Windows edition
[...]
INFO listening vsock://00000400-FACB-11E6-BD58-64006A7986D3
INFO listening \\.\pipe\crc-http
\\.\pipe\crc-http - - [17/Feb/2022:11:40:01 +0530] "GET /api/version HTTP/1.1" 200 74
\\.\pipe\crc-http - - [17/Feb/2022:11:40:20 +0530] "GET /network/services/forwarder/all HTTP/1.1" 200 3
\\.\pipe\crc-http - - [17/Feb/2022:11:40:20 +0530] "POST /network/services/forwarder/expose HTTP/1.1" 200 0
\\.\pipe\crc-http - - [17/Feb/2022:11:40:20 +0530] "POST /network/services/forwarder/expose HTTP/1.1" 200 0
\\.\pipe\crc-http - - [17/Feb/2022:11:40:20 +0530] "POST /network/services/forwarder/expose HTTP/1.1" 500 41
```

**Review by cfergeau — APPROVED** [review]
Looks good to me, maybe wait for more feedback from Anjan on Windows?

**Comment by gbraad** [comment]
@cfergeau Anjan already mentioned it failed due to the SSH key location. is that resolved @praveenkumar ? If so, @anjannath can test again.

**Comment by cfergeau** [comment]
> @cfergeau Anjan already mentioned it failed due to the SSH key location. is that resolved @praveenkumar ? If so, @anjannath can test again.

Anjan tested, reported it's not working, Praveen force-pushed a change which could help with what Anjan was seeing, so I was asking if Anjan can test again.

**Comment by praveenkumar** [comment]
> > @cfergeau Anjan already mentioned it failed due to the SSH key location. is that resolved @praveenkumar ? If so, @anjannath can test again.
> 
> Anjan tested, reported it's not working, Praveen force-pushed a change which could help with what Anjan was seeing, so I was asking if Anjan can test again.

Me and @anjannath tested this PR and in windows also it does able to map the socket but this socket docker client can't consume since it only support named pipe so we need to proxy namedpipe <=> docker sock somehow. 

For windows I created a new tracking issue and from this PR disable the `DOCKER_HOST` from `podman-env` command on windows.

**Comment by praveenkumar** [comment]
Created following issues
- #3018 
- #3019 
- #3020

**Pull request merged**
